### PR TITLE
CLI: Enforce monorepo composer version

### PR DESCRIPTION
**Pull request by sdixon194**
Using monorepo tooling with the wrong composer version can cause composer.lock issues which don't get surfaced until CI checks them. This PR will enforce the composer version being used with the CLI to make sure devs are using the version we need to prevent broken PRs. This PR also lays some groundwork in case we want to do additional kinds of checks in the future.

Check seems performant enough to run on every usage of the CLI, but if it's a problem, we could change when it's used. 
#### Changes proposed in this Pull Request:

Checks the dev environment's composer version against the monorepo required version before running any `jetpack` CLI command, along with instructions on how to fix. *

#### Jetpack product discussion

[N/a](p1649860244940259-slack-C01U2KGS2PQ)
#### Does this pull request change what data or activity we track or use?

No
#### Testing instructions:

* Pull down this PR
* Run `composer self-update` to update composer to the latest version (which isn't supported)
* Try running a CLI command, like `jetpack cli status`
* The command should fail and tell you how to fix your composer version.
* Run the fix suggested.
* Try the command again, it should work fine.*

### CODE REVIEW

File: tools/cli/helpers/checkEnvironment.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * Returns the current dev environment's composer version.
+ *
+ * @returns {string} - the composer version of the current dev environment.
+ */
+export async function getComposerVersion() {
+	const composerString = await child_process
+		.spawnSync( 'composer', [ '--version' ] )
+		.stdout.toString()
+		.trim();
+	const composerVersion = composerString.match( /\d+.\d+.\d+/ );
```
Comment (anomiex) [inline tools/cli/helpers/checkEnvironment.js:29]:
```suggestion
	const composerVersion = composerString.match( /\d+\.\d+\.\d+/ );
```

File: tools/cli/helpers/checkEnvironment.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	return composerVersion[ 0 ];
+}
+
+/**
+ * Compares composer versions and exit if it doesn't match.
+ */
+export async function compareComposerVersion() {
+	const currentComposerVersion = await getComposerVersion();
+	const monorepoComposerVersion = getVersions().COMPOSER_VERSION;
+	if ( currentComposerVersion !== monorepoComposerVersion ) {
```
Comment (anomiex) [inline tools/cli/helpers/checkEnvironment.js:39]: Patch-level differences are ok, we just need to check the major.minor.

We've already got the `semver` package as a dependency, we might just use that to do the comparison. `semver.satisfies( currentComposerVersion, '~' + getVersions().COMPOSER_VERSION )`.

**Review by anomiex — COMMENTED** [review]
If we're going to start failing for e.g. 2.2.5 when versions.sh says 2.2.6, we should update `tools/check-development-environment.sh` too.
```diff
diff --git a/tools/check-development-environment.sh b/tools/check-development-environment.sh
index 258b45ca45..9fe5f03139 100755
--- a/tools/check-development-environment.sh
+++ b/tools/check-development-environment.sh
@@ -279,7 +279,7 @@ if [[ -z "$BIN" ]]; then
 else
 	VER="$(composer --version 2>/dev/null | sed -n -E 's/^Composer( version)? ([0-9]+\.[0-9]+\.[0-9a-zA-Z.-]+) [0-9]{4}-[0-9]{2}-[0-9]{2} [0-9]{2}:[0-9]{2}:[0-9]{2}.*/\2/p')"
 	VX="$(sed -E 's/^([0-9]+\.[0-9]+)\..*/\1/' <<<"$COMPOSER_VERSION")"
-	version_range 'Composer' "$BIN" 'composer' "$VER" "$VX.0" "$COMPOSER_VERSION" "$VX.9999999" true
+	version_range 'Composer' "$BIN" 'composer' "$VER" "$COMPOSER_VERSION" "$COMPOSER_VERSION" "$VX.9999999" true
 fi
 
 checking 'Required extensions are installed'
```

File: tools/cli/helpers/checkEnvironment.js (hunk 2 above)
Comment (sdixon194) [reply tools/cli/helpers/checkEnvironment.js]: [5127afd](https://github.com/Automattic/jetpack/pull/23937/commits/5127afd0d4a0500d9f849bfb993ef3eeab547c8d)

**Review by kraftbj — APPROVED** [review]
Tried `composer self-update` for 2.3.x, `2.2.5` for a disallowed old version, `2.2.6` for the specified and `2.2.7` for a newer patch version.

Both the CLI and `tools/check-development-environment.sh` worked as expected.

The only hesitation is we're saying to install 2.2.6 while 2.2.9 is out and acceptable. But that's not enough of a blocker to stop this from landing given we have folks running into confusion on 2.3. already.

**Pull request merged**